src/Neo/Network/P2P/Payloads/Signer.cs

```diff
@@ -192,14 +192,14 @@ void IInteroperable.FromStackItem(VM.Types.StackItem stackItem)
 
         VM.Types.StackItem IInteroperable.ToStackItem(ReferenceCounter referenceCounter)
         {
-            return new VM.Types.Array(referenceCounter, new VM.Types.StackItem[]
-            {
+            return new VM.Types.Array(referenceCounter,
+            [
                 Account.ToArray(),
                 (byte)Scopes,
-                new VM.Types.Array(referenceCounter, AllowedContracts.Select(u => new VM.Types.ByteString(u.ToArray()))),
-                new VM.Types.Array(referenceCounter, AllowedGroups.Select(u => new VM.Types.ByteString(u.ToArray()))),
-                new VM.Types.Array(referenceCounter, Rules.Select(u => u.ToStackItem(referenceCounter)))
-            });
+                Scopes.HasFlag(WitnessScope.CustomContracts) ? new VM.Types.Array(referenceCounter, AllowedContracts.Select(u => new VM.Types.ByteString(u.ToArray()))) : VM.Types.StackItem.Null,
+                Scopes.HasFlag(WitnessScope.CustomGroups) ? new VM.Types.Array(referenceCounter, AllowedGroups.Select(u => new VM.Types.ByteString(u.ToArray()))) : VM.Types.StackItem.Null,
+                Scopes.HasFlag(WitnessScope.WitnessRules) ? new VM.Types.Array(referenceCounter, Rules.Select(u => u.ToStackItem(referenceCounter))) : VM.Types.StackItem.Null
```

Comment:
Please, push an empty array for the `null` case. It'd be more compatible because we get an empty array for real transactions that have no contracts/groups/rules.